zbus/src/error.rs

```diff
@@ -199,6 +199,12 @@ impl From<nix::Error> for Error {
     }
 }
 
+impl From<crate::address::Error> for Error {
+    fn from(val: crate::address::Error) -> Self {
+        Error::Address(val.to_string())
```

Comment:
What's the point of having a separate Error with variants if we're going to convert it into a string and shove it in a generic error? As you recall, we moved from separate module errors to one error type per crate. Let's merge `address::Error` into `Error` please.

Reply:
It was to make ::address standalone, but does it make sense anymore, maybe not.. I don't know how we could trim zbus to only expose ::address if desired.

Reply:
> It was to make ::address standalone

I realize that.

> but does it make sense anymore

I don't think so, unless you find a way for address to be used standalone but for me the more important question would always be that if we'd have any users of this API and if so, how many. Given that `dbus-addr` crate didn't find any users, I'm very doubtful. Maybe we'll find 1 or 2 user but do we make efforts for some extremely niche use cases? 

I suggested to make use of cargo features. Even though that would involve annoying the hell out of tokio users, who disable the default features. Typical users having to enable the main functionality of the crate, just for niche use case, isn't the best idea. Still, if you think it's worth it, I won't stop you from it.

If you don't decide to work on it soon, then we should merge the error types.

Reply:
also desirable is to compile-time check addresses..

how would you like to merge errors? drop all the variants of address::Error and replace them with the generic Error::Address ?

Reply:
> how would you like to merge errors? drop all the variants of address::Error and replace them with the generic Error::Address ?

I think so, yes. I seriously doubt user can programatically do anything different based on the exact variant and recover for it using the same address. Also, this is what is happening currently after this PR anyway when user uses `connection::Builder::address`.

Reply:
While it may not be handled programmatically, a precise error can allow for better error feedback returned to the user. It's internal anyway, is it really worth to change now ? 

I think it's a shame to conflate the various errors that can be exposed to a flat String. Having a single Address error, imho we lose some robustness. We trade a well set of meaningful errors for a wildcard..

Why don't you do this ? It makes me sad to have to disagree with your opinions...

Reply:
> While it may not be handled programmatically, a precise error can allow for better error feedback returned to the user.

That's what the description string is for. If they print/debug the error, they'll see the string that explains the exact problem.

> I think it's a shame to conflate the various errors that can be exposed to a flat String. Having a single Address error, imho we lose some robustness. We trade a well set of meaningful errors for a wildcard..

Let's be more specific. Can you give an example of a real-world situation where different variants would help the application programmer? I have to remind you again that most users will not even use this API directly, let alone parse the variants.

BTW, you missed a very important point:

>  Also, this is what is happening currently after this PR anyway when user uses connection::Builder::address.

So if you really want a separate error, at least you should then go the whole mile and use this type in the `zbus::Error::Address`. I don't think we should though for reasons mentioned in this comment and before.

> Why don't you do this ? It makes me sad to have to disagree with your opinions...

Same here but I don't see any connection. In the end it's the same result and you promised to maintain this code and this is very much part of the maintenance IMO. I still think that these changes unnecessarily complicated the API with no **real** benefits but I accepted it because you wouldn't let it go and since it will not affect most users in any way.

Reply:
We should keep address::Error separate, because it carries semantics that can be used by API users. They can provide relevant information and feedback when parsing an address. And for testing alone, it's superior to have an enum to match against, rather than a string.

Reply:
I know the theoretical benefits. I was asking for more practical ones. I already explained in detail above why I think we should merge it. If you want to argue, please address all my specific points above. Otherwise, it doesn't take the discussion anywhere.

Reply:
"They can provide relevant information and feedback when parsing an address" think a tool that validates the dbus address, it could be a management tool or for example a middlelayer like libvirt which manipulates and takes dbus configuration from users. You can provide different feedback based on the different errors, provide link to reference documentation, examples, suggestions to fix etc... & I don't need to detail the testing scenario.

This is not something you can reliably do if all you have is a single Error::Address(String) variant.

address::Error has to be public already, because we use this type for returning errors for various conversion traits which are publics. Potentially, users may have to deal with it if they do such conversions themself.

I'll modify the code to expose the address::Error to zbus::Error.

Reply:
> "They can provide relevant information and feedback when parsing an address" think a tool that validates the dbus address, it could be a low-level tool, or a management tool or for example a middlelayer like libvirt which manipulates and takes dbus configuration from users. You can provide different feedback based on the different errors, provide link to reference documentation, examples, suggestions to fix etc... & I don't need to detail the testing scenario.

Now we're talking. Actual practical examples. :+1:

> I'll modify the code to expose the address::Error to zbus::Error

Yes please.

Reply:
done in #1050